api/src/services/QueueManager.ts

```diff
@@ -35,20 +35,33 @@ class QueueManager {
         // that is already awaiting indexing.
         const q = new Queue(this.defaultQueueName);
         const jobs = await q.getJobs("wait");
-        let lastPidAction = null;
-        for (let i = 0; i < jobs.length; i++) {
-            if (jobs[i].name === "index" && jobs[i].data.pid === pid) {
-                lastPidAction = jobs[i].data.action;
-                break;
-            }
-        }
-        if (action === lastPidAction) {
+        const queueJob = { pid, action };
+        if (this.isAlreadyAwaitingAction(jobs, "metadata", queueJob)) {
             console.log("Skipping queue; " + pid + " is already awaiting " + action + ".");
         } else {
             await q.add("index", { pid, action });
         }
         q.close();
     }
+
+    isAlreadyAwaitingAction(jobs: Array<Job>, name: string, { pid, action }: { pid: string; action: string }): boolean {
+        const matchingJob = jobs.find((job) => {
+            return job.name === name && job.data.pid === pid;
```

Comment:
I think this may be an oversimplification -- isn't this going to find the first match in the queue? If so, I think what we really want is to find the last match in the queue. In other words, we might have a queue like:

index, delete, index, delete, index, delete, index, delete

If the next action to be queued is index, we want to add it to the end, since after we have deleted a record, we need to bring it back if it is revived. But if we pull "index" off the front of the queue and then drop the job because it matches, the record will incorrectly stay deleted.

Apologies if I'm misunderstanding the code and this is already accounted for.

Reply:
...in any case, I guess this makes for a more sophisticated test case that's worth adding!

Reply:
I misunderstood that check.

Reply:
Okay. the original code looks at the first instance. Maybe we should discuss this?

Reply:
Ha, you're right, I think that may be a bug on my part, or else I'm misremembering the order of the queue. If it's newest-to-oldest, then it's fine... but if it's oldest-to-newest, then I think it's wrong.